keras_cv/models/object_detection/faster_rcnn.py

```diff
@@ -346,81 +345,165 @@ def _call_rpn(self, images, training=None):
+        box_pred, cls_pred = self._call_rcnn(rois, feature_map)
+        y_true = {
+            "rpn_box": rpn_box_targets,
+            "rpn_cls": rpn_cls_targets,
+            "box": box_targets,
+            "cls": cls_targets,
+        }
+        y_pred = {
+            "rpn_box": rpn_box_pred,
+            "rpn_cls": rpn_cls_pred,
+            "box": box_pred,
+            "cls": cls_pred,
+        }
+        weights = {
+            "rpn_box": rpn_box_weights,
+            "rpn_cls": rpn_cls_weights,
+            "box": box_weights,
+            "cls": cls_weights,
+        }
+        return super().compute_loss(
+            x=images, y=y_true, y_pred=y_pred, sample_weight=weights
         )
-        # [BS, H*W*K, 4], [BS, H*W*K, num_classes + 1]
-        rcnn_box_pred, rcnn_cls_pred = self.rcnn_head(feature_map)
-        res = {}
-        if training:
-            res["rcnn_box_pred"] = rcnn_box_pred
-            res["rcnn_cls_pred"] = rcnn_cls_pred
-            res["rpn_box_pred"] = tf.concat(tf.nest.flatten(rpn_boxes), axis=1)
-            res["rpn_cls_pred"] = tf.concat(tf.nest.flatten(rpn_scores), axis=1)
-            res["rcnn_box_targets"] = rcnn_box_targets
-            res["rcnn_box_weights"] = rcnn_box_weights
-            res["rcnn_cls_targets"] = rcnn_cls_targets
-            res["rcnn_cls_weights"] = rcnn_cls_weights
-        else:
-            # now it will be on "center_yxhw" format
-            rcnn_box_pred = _decode_deltas_to_boxes(
-                anchors=rois,
-                boxes_delta=rcnn_box_pred,
-                anchor_format="yxyx",
-                box_format=self.bounding_box_format,
-                variance=[0.1, 0.1, 0.2, 0.2],
-            )
-            res["rcnn_box_pred"] = rcnn_box_pred
-            res["rcnn_cls_pred"] = rcnn_cls_pred
 
-        return res
+    def train_step(self, data):
+        images = data["images"]
+        gt_boxes = data["gt_boxes"]
+        gt_classes = data["gt_classes"]
+        with tf.GradientTape() as tape:
+            total_loss = self.compute_loss(images, gt_boxes, gt_classes, training=True)
+            reg_losses = []
+            if self.weight_decay:
+                for var in self.trainable_variables:
+                    if "bn" not in var.name:
+                        reg_losses.append(self.weight_decay * tf.nn.l2_loss(var))
+                l2_loss = tf.math.add_n(reg_losses)
+            total_loss += l2_loss
+        self.optimizer.minimize(total_loss, self.trainable_variables, tape=tape)
+        return self.compute_metrics(images, {}, {}, sample_weight={})
+
+    def test_step(self, data):
+        images = data["images"]
+        gt_boxes = data["gt_boxes"]
+        gt_classes = data["gt_classes"]
+        self.compute_loss(images, gt_boxes, gt_classes, training=False)
+        return self.compute_metrics(images, {}, {}, sample_weight={})
+
+
+def _validate_loss(loss, loss_name):
```

Comment:
Should we also validate that, for losses which require a box format, the format matches the format of the rest of the model?

Reply:
loss is independent to box format, IIUC. So users should be able to change between formats and still use the same loss. However they would have to scale differently

Reply:
Is IOU loss not relevant for this model? (IIRC that requires a box format?)

Reply:
There seem to be inconsistent losses: 1) iou loss that require a format; 2) smoothl1 loss that doesn't require a format.
In the long run, we should probably by fixing it so that it's format independent. Why? Because there are some weird format (and actually mostly used compared to other formats) such as `tx` `ty` `tw` `th` which we wouldn't be able to scale up by supporting all of them, and clearly we don't want to block users if they want to convert their format to a new one before passing to loss function.